### PR TITLE
CDAP-14929 Define all tables before starting services in Cloud Runtime.

### PR DESCRIPTION
**Pull request by anwar6953**
https://issues.cask.co/browse/CDAP-14929

Cloud integration tests: https://builds.cask.co/browse/IT-CLD-208
Unit tests: https://builds.cask.co/browse/CDAP-DUT6836-4

### CODE REVIEW

**Review by sreevatsanraman — APPROVED** [review]

**Comment by anwar6953** [comment]
@anew I know you suggested a change such as https://github.com/cdapio/cdap/commit/a0d8c8e1e3e89a3cbcf2ddcc2ce97d130f9078f3, but that wasn't necessary. Rather, with it, the fix doesn't work.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java [hunk 1]
```diff
@@ -507,6 +511,13 @@ private ProgramOptions createProgramOptions(File programOptionsFile) throws IOEx
         addOnPremiseServices(injector, programOptions, metricsCollectionService, services);
         break;
       case ISOLATED:
+        try {
+          // Define all StructuredTable before starting any services that need StructuredTable
+          StoreDefinition.createAllTables(injector.getInstance(StructuredTableAdmin.class),
+                  injector.getInstance(StructuredTableRegistry.class));
+        } catch (IOException | TableAlreadyExistsException e) {
+          throw new RuntimeException("Unable to create the system tables.", e);
+        }
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java:520]: I thought we only need to create the tables used by dataset service? Why create all of them?

Comment (anwar6953) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java:520]: That would probably be just these 2, then?

```java
    if (overWrite || tableAdmin.getSpecification(DatasetInstanceStore.DATASET_INSTANCES) == null) {
      DatasetInstanceStore.createTables(tableAdmin);
    }
    if (overWrite || tableAdmin.getSpecification(DatasetTypeStore.DATASET_TYPES) == null) {
      DatasetTypeStore.createTables(tableAdmin);
    }
```

Comment (anew) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractProgramTwillRunnable.java:520]: yes

**Review by anew — APPROVED** [review]

**Pull request merged**